src/random/cpu/choice.cc

```diff
@@ -26,73 +26,90 @@ IdxType RandomEngine::Choice(FloatArray prob) {
+  for (IdxType i = 0; i < num; ++i) out[i] = sampler->Draw();
   delete sampler;
 }
 
-template void RandomEngine::Choice<int32_t, float>(
-    int32_t num, FloatArray prob, int32_t* out, bool replace);
-template void RandomEngine::Choice<int64_t, float>(
-    int64_t num, FloatArray prob, int64_t* out, bool replace);
-template void RandomEngine::Choice<int32_t, double>(
-    int32_t num, FloatArray prob, int32_t* out, bool replace);
-template void RandomEngine::Choice<int64_t, double>(
-    int64_t num, FloatArray prob, int64_t* out, bool replace);
+template void RandomEngine::Choice<int32_t, float>(int32_t num, FloatArray prob,
+                                                   int32_t* out, bool replace);
+template void RandomEngine::Choice<int64_t, float>(int64_t num, FloatArray prob,
+                                                   int64_t* out, bool replace);
+template void RandomEngine::Choice<int32_t, double>(int32_t num,
+                                                    FloatArray prob,
+                                                    int32_t* out, bool replace);
+template void RandomEngine::Choice<int64_t, double>(int64_t num,
+                                                    FloatArray prob,
+                                                    int64_t* out, bool replace);
 
 template <typename IdxType>
-void RandomEngine::UniformChoice(IdxType num, IdxType population, IdxType* out, bool replace) {
+void RandomEngine::UniformChoice(IdxType num, IdxType population, IdxType* out,
+                                 bool replace) {
   if (!replace)
-    CHECK_LE(num, population) << "Cannot take more sample than population when 'replace=false'";
+    CHECK_LE(num, population)
+      << "Cannot take more sample than population when 'replace=false'";
   if (replace) {
-    for (IdxType i = 0; i < num; ++i)
-      out[i] = RandInt(population);
+    for (IdxType i = 0; i < num; ++i) out[i] = RandInt(population);
   } else {
-    if (num < population / 10) {  // TODO(minjie): may need a better threshold here
-      // use hash set
-      // In the best scenario, time complexity is O(num), i.e., no conflict.
-      //
-      // Let k be num / population, the expected number of extra sampling steps is roughly
-      // k^2 / (1-k) * population, which means in the worst case scenario,
-      // the time complexity is O(population^2). In practice, we use 1/10 since
-      // std::unordered_set is pretty slow.
-      std::unordered_set<IdxType> selected;
-      while (selected.size() < num) {
-        selected.insert(RandInt(population));
+    if (num <
+        population / 10) {  // TODO(minjie): may need a better threshold here
+      if (num && num < 64) {
+        *out = RandInt(population);
+        auto b = out + 1;
+        auto e = b + num - 1;
+        while (b != e) {
+          *b = RandInt(population);
+          auto it = std::find(out, b, *b);
+          if (it != b) continue;
+          ++b;
+        }
+
+      } else {
```

Comment:
Looks good to me, would you minding mention this PR in the comments?

Reply:
@yzh119 done

Reply:
perfect